### PR TITLE
Start of Morello kernel support

### PR DESCRIPTION
**Pull request by zxombie**

### CODE REVIEW

File: sys/arm64/arm64/trap.c
Comment (arichardson) [inline sys/arm64/arm64/trap.c:302]: I assume there is a follow-up patch to print these as capregs? If not there should be a TODO here.

File: sys/arm64/include/elf.h
Comment (arichardson) [inline sys/arm64/include/elf.h:64]:
```suggestion
#if __SIZEOF_POINTER__ == 8
```

File: sys/arm64/include/frame.h
Comment (arichardson) [inline sys/arm64/include/frame.h:46]: I believe we now define uintcap_t unconditionally (to uint64_t if it isn't supported), so we could drop the else and only ifdef the addition of `tf_ddc`.

File: sys/arm64/include/frame.h
Comment (arichardson) [inline sys/arm64/include/frame.h:58]: why is this needed? the struct is not packed, so tf_x will be aligned correctly by the compiler.

File: sys/arm64/include/frame.h
Comment (arichardson) [reply sys/arm64/include/frame.h:46]: Unless this is yet another case of unnecessary stdint.h/stddef.h duplication for each architecture... In which case we should just add uintcap_t to an sys/_types.h

File: sys/arm64/include/elf.h
Comment (jrtc27) [reply sys/arm64/include/elf.h:64]: These are not the same thing; e.g. CHERI64 would match the latter not the former. Leave it as it is; it’s also what we use elsewhere.

File: sys/arm64/arm64/trap.c
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:302]: Again, we’re committing workarounds here for things that will be changed (I hope). This is needless churn that polluted the history. We do not need that merged.

Comment (arichardson) [reply sys/arm64/arm64/trap.c:302]: The line needs to be changed again anyway once we print caps, I don't think it matters much whether we touch it once or twice. Especially since we won't be able to compile the kernel without this change unless we turn of -Wformat.

File: sys/arm64/include/frame.h
Comment (zxombie) [reply sys/arm64/include/frame.h:58]: This is to ensure there are no holes in the struct that have an undefined value after a copy.

File: sys/arm64/include/elf.h
Comment (zxombie) [reply sys/arm64/include/elf.h:64]: Both `mips` and `riscv` use `__ELF_WORD_SIZE`

File: sys/arm64/include/frame.h
Comment (zxombie) [reply sys/arm64/include/frame.h:46]: Where is that typedef? The riscv trapframe defines this in a similar way.

File: sys/arm64/arm64/trap.c
Comment (zxombie) [reply sys/arm64/arm64/trap.c:302]: I can add a TODO so I can get back to it after image activator changes are ready.

File: sys/arm64/include/frame.h
Comment (arichardson) [reply sys/arm64/include/frame.h:46]: sys/sys/_types.h
```c
#if !__has_feature(capabilities)
typedef	__intptr_t	__intcap_t;
typedef	__uintptr_t	__uintcap_t;
#endif
```

sys/sys/_stdint.h (added in 2018, 8a64dd37f2ae722a41fca70383d53c2b0b55c70a)
```c
#ifndef _INTCAP_T_DECLARED
typedef	__intcap_t		intcap_t;
#define	_INTCAP_T_DECLARED
#endif
#ifndef _UINTCAP_T_DECLARED
typedef	__uintcap_t		uintcap_t;
#define	_UINTCAP_T_DECLARED
#endif
```

And that file also defines kintcap_t/kuintcap_t for things that should only be __(u)intcap_t if _KERNEL is defined.

File: sys/arm64/include/elf.h
Comment (arichardson) [reply sys/arm64/include/elf.h:64]: Yeah I don't really care, I just prefer positive checks rather than negative ones.
`#if __ELF_WORD_SIZE == 64 && __SIZEOF_POINTER__ == 8` would also work for the cheri64 purecap case.

Comment (jrtc27) [reply sys/arm64/include/elf.h:64]: In practice the sets of things that matches and the original code matches are the same, but only the original code says what it _means_. Your positive form makes you have to stop and think more IMO, and for anyone who doesn't know what CHERI is they'd be very confused, yet the mention of CHERI and capabilities in this version tells them what's actually going on.

Comment (arichardson) [reply sys/arm64/include/elf.h:64]: I was not sure whether __ELF_WORD_SIZE always matches __SIZEOF_POINTER, but it appears that e.g. the x32 ABI uses ELF32, so there are probably no ABIs other than CHERI where `__ELF_WORD_SIZE != __SIZEOF_POINTER__ * _CHAR_BIT`.

I really don't care, I just personally find __SIZEOF_POINTER__ easier to understand.

File: sys/arm64/arm64/locore.S
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:626]: I still believe we should always be using ATTR_DEFAULT_LOW and ATTR_DEFAULT_HIGH here

File: sys/arm64/arm64/swtch.S
Comment (jrtc27) [inline sys/arm64/arm64/swtch.S:232]:
```suggestion
	/* Restore the registers other than x0 and x1 */
#if __has_feature(capabilities)
	ldp	c2, c3, [sp, #TF_X + 2 * 16]
	ldp	c4, c5, [sp, #TF_X + 4 * 16]
	ldp	c6, c7, [sp, #TF_X + 6 * 16]
	ldp	c8, c9, [sp, #TF_X + 8 * 16]
	ldp	c10, c11, [sp, #TF_X + 10 * 16]
	ldp	c12, c13, [sp, #TF_X + 12 * 16]
	ldp	c14, c15, [sp, #TF_X + 14 * 16]
	ldp	c16, c17, [sp, #TF_X + 16 * 16]
	ldr	     c19, [sp, #TF_X + 19 * 16]
	ldp	c20, c21, [sp, #TF_X + 20 * 16]
	ldp	c22, c23, [sp, #TF_X + 22 * 16]
	ldp	c24, c25, [sp, #TF_X + 24 * 16]
	ldp	c26, c27, [sp, #TF_X + 26 * 16]
	ldp	c28, c29, [sp, #TF_X + 28 * 16]
#else
```

to match the other places, or systematically duplicate the comments on each side of the #if and change x to c.

File: sys/arm64/arm64/trap.c
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:306]: So either:
(a) turn off -Wformat as an early commit and then make this line print the capability
(b) use a version of the compiler with Alex's cherry-picked patches
It's not hard to avoid the problem here.

File: sys/arm64/arm64/trap.c
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:310]: As above

File: sys/arm64/include/atomic.h
Comment (jrtc27) [inline sys/arm64/include/atomic.h:78]: I'd still prefer a better name for this, hmm.

File: sys/arm64/include/cheri.h
Comment (jrtc27) [inline sys/arm64/include/cheri.h:38]: We have these casts for mips and riscv but they're probably not needed now with polymorphic builtins? sys/sys/systm.h uses each once with ___USER_CFROMPTR, and sys/cheri/cheri_syscall.c does a `cheri_getperm(__USER_PCC)`, both of which should work fine now? I'll submit a PR to remove those from mips and riscv and run it through CI.

File: sys/arm64/include/pte.h
Comment (jrtc27) [inline sys/arm64/include/pte.h:47]: What are the other two states reserved for; can we give them names here?

File: sys/arm64/include/reg.h
Comment (jrtc27) [inline sys/arm64/include/reg.h:83]:
```suggestion
	uintcap_t c[30];
	uintcap_t clr;
	uintcap_t csp;
	uintcap_t celr;
	uintcap_t ddc;
```

**Review by jrtc27 — CHANGES_REQUESTED** [review]

File: sys/arm64/arm64/swtch.S
Comment (jrtc27) [reply sys/arm64/arm64/swtch.S:232]: I don't know why GitHub is rendering the diff in such a strange way, it just moves the line from the bottom to the top...

File: sys/arm64/arm64/trap.c
Comment (arichardson) [reply sys/arm64/arm64/trap.c:306]: I really don't like option a). This is a genuine bug. It happens to work correctly in this case since the vararg happens to be passed in registers, but if this was a printf with more arguments we would print the wrong value for later args.

File: sys/arm64/arm64/trap.c
Comment (arichardson) [inline sys/arm64/arm64/trap.c:308]:
```suggestion
#if __has_feature(capabilities)
	printf("  sp: " _CHERI_PRINTF_CAP_FMT "\n",
	     _CHERI_PRINTF_CAP_ARG((void * __capability)frame->tf_sp));
#else
	printf("  sp: %16lx\n", (uint64_t)frame->tf_sp);
#endif
```
Or the RISC-V version:
```c
#if __has_feature(capabilities)
/* This cannot use _CHERI_PRINTF_CAP_ARG due to the cast for purecap. */
#ifdef __CHERI_PURE_CAPABILITY__
#define	PRINT_REG_ARG(value)	((void * __capability)(value))
#else
#define	PRINT_REG_ARG(value)	(&(value))
#endif
#define PRINT_REG(name, value)					\
	printf(name " = " _CHERI_PRINTF_CAP_FMT "\n",		\
	    PRINT_REG_ARG(value));
#define PRINT_REG_N(name, n, array)				\
	printf(name "[%d] = " _CHERI_PRINTF_CAP_FMT "\n", n,	\
	    PRINT_REG_ARG((array)[n]));
#else
#define PRINT_REG(name, value)	printf(name " = 0x%016lx\n", value)
#define PRINT_REG_N(name, n, array)	\
	printf(name "[%d] = 0x%016lx\n", n, (array)[n])
#endif
```

File: sys/arm64/arm64/trap.c
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:306]: There's no bug here? You'd make the format string be a capability for CHERI kernels and the argument be a capability, and just turn off -Wformat because the compiler doesn't yet know the format string is valid?

File: sys/arm64/include/cheri.h
Comment (jrtc27) [reply sys/arm64/include/cheri.h:38]: Never mind, tried in #707 but turns out they are ultimately needed.

File: sys/arm64/arm64/trap.c
Comment (arichardson) [reply sys/arm64/arm64/trap.c:306]: tf_x is a `uintcap_t` in the `__has_feature(capabilities)` case, so -Wformat complains.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:306]: Yes, so use `%#.16lp` in that case, ie _CHERI_PRINTF_CAP_FMT.

Comment (arichardson) [reply sys/arm64/arm64/trap.c:306]: I missed the "then make this line print the capability" part of option a). I thought you were suggesting we simply disable warnings until printing capregs has been implemented.

File: sys/arm64/include/frame.h
Comment (zxombie) [reply sys/arm64/include/frame.h:46]: Can I wait until after this is merged for this given it's similar to riscv?

File: sys/arm64/arm64/locore.S
Comment (zxombie) [reply sys/arm64/arm64/locore.S:626]: And I've already explained why I prefer doing it this way.

Comment (jrtc27) [reply sys/arm64/arm64/locore.S:626]: It doesn't reduce the diff though; every assembly use of ATTR_DEFAULT still needs to be touched regardless, so why not do it the more-general way?

Comment (zxombie) [reply sys/arm64/arm64/locore.S:626]: There are 2 cases. I don't expect any more to be added as they are only needed to bootstrap the pagetables enough to get into C.

Comment (jrtc27) [reply sys/arm64/arm64/locore.S:626]: So it's easy to fix now then.

File: sys/arm64/include/frame.h
Comment (arichardson) [reply sys/arm64/include/frame.h:46]: Reducing ifdefs is always nice, but I don't really mind either way.

File: sys/arm64/arm64/locore.S
Comment (zxombie) [reply sys/arm64/arm64/locore.S:626]: There is nothing to fix

Comment (jrtc27) [reply sys/arm64/arm64/locore.S:626]:
```suggestion
#if ATTR_DEFAULT_HIGH != 0
	orr	x12, x12, #(ATTR_DEFAULT_HIGH)
#endif
	orr	x12, x12, #(ATTR_DEFAULT_LOW)
```
is shorter, is clearer / more readable, and keeps more common code between Morello and plain kernels.

File: sys/arm64/arm64/trap.c
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:298]: Given the kernel is missing large amounts of features I would not include this diff and just leave it broken to fix in a later commit.

Comment (arichardson) [reply sys/arm64/arm64/trap.c:298]: It won't compile if we do that. I preferred casts+TODO. We have to ifdef it out if we don't change the format strings.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:298]: Or pragmas to fix the -Wformat warnings if that's the issue?

Comment (arichardson) [reply sys/arm64/arm64/trap.c:298]: As long as there is a diff to upstream that reminds us to implement this, I don't mind what we do for now.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:298]: Well I would fix it to have the right code, and only use the pragma to deal with outdated toolchain issues.

File: sys/arm64/arm64/locore.S
Comment (brooksdavis) [reply sys/arm64/arm64/locore.S:626]: I'm ok with the from @zxombie proposes unless it makes sense to upstream ATTR_DEFAULT_LOW and ATTR_DEFAULT_LOW in the short term.

Comment (jrtc27) [reply sys/arm64/arm64/locore.S:626]: If going with that form I'd prefer the condition be `#if __has_feature(capabilities)`, since currently it's a bit weird that the HIGH use is guarded by there being a LOW.

File: sys/arm64/arm64/trap.c
Comment (arichardson) [reply sys/arm64/arm64/trap.c:298]: Yes that would obviously be better (fixes have been backported so no need for pragmas), but I don't think that needs to be part of this PR.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:298]: See, I really don't like this break-a-function-just-so-it-compiles approach, it pollutes the history. It's fine for development, but the cleaned-up history that we merge in should not include such hacks, and just be a clean series of commits that get you from no CHERI to CHERI. Why can the single commit to fix the printing here not be a part of this PR? It should be independent from everything else.

Comment (arichardson) [reply sys/arm64/arm64/trap.c:298]: That's why I preferred the cast+TODO version since it would have printed useful output and compiled correctly.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:298]: Yeah, not my preferred choice, but would be adequate in my eyes, since it's not _wrong_ in the sense that it's broken, insecure or less functional than baseline aarch64 (like some of the other things I've objected to before), just less helpful than it could be.

Comment (jrtc27) [reply sys/arm64/arm64/trap.c:298]: ... but also we don't have copyincap/copyoutcap yet so the kernel will fail to link, and thus talking about whether this file compiles or not is a bit pointless.

Comment (zxombie) [reply sys/arm64/arm64/trap.c:298]: I'm happy to switch back to casts + TODO.

File: sys/arm64/arm64/locore.S
Comment (zxombie) [reply sys/arm64/arm64/locore.S:626]: I can change it to `#if __has_feature(capabilities)`

Comment (zxombie) [reply sys/arm64/arm64/locore.S:626]: `ATTR_DEFAULT_LOW` and `ATTR_DEFAULT_HIGH` don't make sense upstream as `ATTR_DEFAULT` on non-morello arm64 fits within the operand of a singe `orr` instruction.

File: sys/arm64/arm64/locore.S
Comment (jrtc27) [inline sys/arm64/arm64/locore.S:664]: This one didn't get changed.

Comment (zxombie) [reply sys/arm64/arm64/locore.S]: Fixed

File: sys/arm64/include/armreg.h
Comment (jrtc27) [inline sys/arm64/include/armreg.h:86]: Guard with __has_feature(capabilities)

File: sys/arm64/include/hypervisor.h
Comment (jrtc27) [inline sys/arm64/include/hypervisor.h:47]: Condition based on __has_feature(capabilities)? Certainly CPTR_TC should be guarded, and CPTR_RES1 is used by locore so I assume it needs the old value for non-Morello, and perhaps it also wants CPTR_TC for Morello too.

File: sys/arm64/include/pte.h
Comment (jrtc27) [inline sys/arm64/include/pte.h:50]: Guard with __has_feature(capabilities)

File: sys/arm64/include/hypervisor.h
Comment (jrtc27) [inline sys/arm64/include/hypervisor.h:51]: These aren't Morello-related.

**Comment by jrtc27** [comment]
Please fold that commit into the earlier relevant one(s).

File: sys/arm64/arm64/exception_morello.S
Comment (jrtc27) [inline sys/arm64/arm64/exception_morello.S:2]:
```suggestion
 * Copyright (c) 2014, 2020 Andrew Turner
```

File: sys/arm64/include/atomic.h
Comment (jrtc27) [reply sys/arm64/include/atomic.h:78]: Maybe `_ATOMIC_PTR_CNSTR` (or if one more letter is ok, `_ATOMIC_PTR_CONSTR`)?

Comment (arichardson) [reply sys/arm64/include/atomic.h:78]: Adding that "R" sounds good since it isn't obvious that the CONST means "inlineasm constraint" rather than "const pointer" without looking at the uses of this macro.

File: sys/arm64/include/armreg.h
Comment (brooksdavis) [inline sys/arm64/include/armreg.h:87]: Indenting seems a bit odd (but that seems indemic to this file so probably not a blocker.)

I do wonder if defining an upstreamable `CPACR_CEN_SHIFT` and avoiding the large ifdef would be cleaner.

**Review by brooksdavis — APPROVED** [review]

File: sys/arm64/include/hypervisor.h
Comment (jrtc27) [reply sys/arm64/include/hypervisor.h:48]: Does locore do the right thing or not with the reduced CPTR_RES1?

Comment (jrtc27) [reply sys/arm64/include/hypervisor.h:48]: (if the answer is "I don't know", that's fine, but you should add a suitable XXXAT either here or there)

File: sys/arm64/include/armreg.h
Comment (zxombie) [reply sys/arm64/include/armreg.h:87]: It should be the standard indentation for this file, the field values have an extra space to help find different fields visually.

File: sys/arm64/include/hypervisor.h
Comment (zxombie) [reply sys/arm64/include/hypervisor.h:48]: We can use capabilities at EL1 and below without raising an exception when the `CPTR_TC` bit is clear.

Comment (jrtc27) [reply sys/arm64/include/hypervisor.h:48]: Which means what for my question?

Comment (zxombie) [reply sys/arm64/include/hypervisor.h:48]: On the current version of the patch `CPTR_RES1` has the `CPTR_TC` field clear in a Morello kernel and set in a legacy kernel so should be correct.

Comment (jrtc27) [reply sys/arm64/include/hypervisor.h:48]: So `CPTR_RES1` is "trap if the guest tries to use any feature we don't know about", i.e. since a CHERI-aware kernel knows about capabilities there is no need to make them trap in the guest any more?

Comment (zxombie) [reply sys/arm64/include/hypervisor.h:48]: It is the fields that are marked as `RES1`, i.e. unknown fields that you should set to `1`. In this case bit 9 was `RES1` in armv8, and is now `TC` in Morello. The `TC` field is set to `1` when you want to trap capabilities at EL2 so a CHERI-aware kernel that understands capabilities shouldn't trap. The guest is the same kernel configuring the `CPTR_EL2` register. It starts in EL2, then drops to EL1 after configuring the hypervisor registers.

**Pull request merged**
